### PR TITLE
New benchmark derived from otel-launcher-go

### PR DESCRIPTION
**Pull request by jmacd**
This is a near-identical copy of the benchmark used in developing the [Lighstep metrics SDK](https://github.com/lightstep/otel-launcher-go/tree/main/lightstep/sdk/metric).

This measures end-to-end performance of the synchronous instrument implementation using Counters. This should help settle the question in https://github.com/open-telemetry/opentelemetry-go/issues/3199.

The current numbers:

```
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/metric
BenchmarkCounterAddNoAttrs-10                 	51228051	        22.61 ns/op	       0 B/op	       0 allocs/op
BenchmarkCounterAddOneAttr-10                 	 7272368	       164.0 ns/op	     152 B/op	       3 allocs/op
BenchmarkCounterAddOneInvalidAttr-10          	 4732094	       254.3 ns/op	     280 B/op	       3 allocs/op
BenchmarkCounterAddManyAttrs-10               	 2170496	       638.3 ns/op	     301 B/op	       3 allocs/op
BenchmarkCounterAddManyInvalidAttrs-10        	 1676473	       658.9 ns/op	     378 B/op	       3 allocs/op
BenchmarkCounterAddManyFilteredAttrs-10       	 1610043	       676.1 ns/op	     381 B/op	       3 allocs/op
BenchmarkCounterCollectOneAttr-10             	 2484734	       475.9 ns/op	     456 B/op	       8 allocs/op
BenchmarkCounterCollectTenAttrs-10            	  507835	      2352 ns/op	    2512 B/op	      35 allocs/op
BenchmarkCounterCollectTenAttrsTenTimes-10    	   51271	     23737 ns/op	   25120 B/op	     350 allocs/op
```

### CODE REVIEW

**Comment by jmacd** [comment]
Although these benchmarks allow comparing the two SDKs, that wasn't my goal. Here is the corresponding output for otel-launcher-go:

```
goos: darwin
goarch: arm64
pkg: github.com/lightstep/otel-launcher-go/lightstep/sdk/metric
BenchmarkCounterAddNoAttrs-10                 	30648045	        37.15 ns/op	       0 B/op	       0 allocs/op
BenchmarkCounterAddOneAttr-10                 	12716595	        93.52 ns/op	      64 B/op	       1 allocs/op
BenchmarkCounterAddOneInvalidAttr-10          	 8578038	       140.8 ns/op	     128 B/op	       1 allocs/op
BenchmarkCounterAddManyAttrs-10               	 1000000	      1157 ns/op	     569 B/op	       6 allocs/op
BenchmarkCounterAddManyInvalidAttrs-10        	  743876	      1799 ns/op	    1092 B/op	      10 allocs/op
BenchmarkCounterAddManyFilteredAttrs-10       	 1000000	      1426 ns/op	     954 B/op	       8 allocs/op
BenchmarkCounterCollectOneAttrNoReuse-10      	 2412259	       493.3 ns/op	     400 B/op	       7 allocs/op
BenchmarkCounterCollectOneAttrWithReuse-10    	 3470206	       345.5 ns/op	     136 B/op	       3 allocs/op
BenchmarkCounterCollectTenAttrs-10            	  663994	      1762 ns/op	     712 B/op	      12 allocs/op
BenchmarkCounterCollectTenAttrsTenTimes-10    	   68974	     17650 ns/op	    7129 B/op	     120 allocs/op
```

See it uses fewer allocations and is somewhat faster when the attribute set is re-used (due to "fingerprinting"), otherwise is somewhat slower due to extra complexity associated with memory management.

**Comment by jmacd** [comment]
Also note @bogdandrutu I could not identify a performance impact of the following diff

```
diff --git a/sdk/metric/view/view.go b/sdk/metric/view/view.go
index c6f20018..963c7240 100644
--- a/sdk/metric/view/view.go
+++ b/sdk/metric/view/view.go
@@ -71,7 +71,7 @@ func New(opts ...Option) (View, error) {
 
 // TransformInstrument will check if an instrument matches this view
 // and will convert it if it does.
-func (v View) TransformInstrument(inst Instrument) (transformed Instrument, match bool) {
+func (v *View) TransformInstrument(inst Instrument) (transformed Instrument, match bool) {
 	if !v.match(inst) {
 		return Instrument{}, false
 	}
@@ -89,7 +89,7 @@ func (v View) TransformInstrument(inst Instrument) (transformed Instrument, matc
 
 // AttributeFilter returns a function that returns only attributes specified by
 // WithFilterAttributes. If no filter was provided nil is returned.
-func (v View) AttributeFilter() func(attribute.Set) attribute.Set {
+func (v *View) AttributeFilter() func(attribute.Set) attribute.Set {
 	if v.filter == nil {
 		return nil
 	}
@@ -99,27 +99,27 @@ func (v View) AttributeFilter() func(attribute.Set) attribute.Set {
 	}
 }
 
-func (v View) matchName(name string) bool {
+func (v *View) matchName(name string) bool {
 	return v.instrumentName == nil || v.instrumentName.MatchString(name)
 }
 
-func (v View) matchScopeName(name string) bool {
+func (v *View) matchScopeName(name string) bool {
 	return v.scope.Name == "" || name == v.scope.Name
 }
 
-func (v View) matchScopeVersion(version string) bool {
+func (v *View) matchScopeVersion(version string) bool {
 	return v.scope.Version == "" || version == v.scope.Version
 }
 
-func (v View) matchScopeSchemaURL(schemaURL string) bool {
+func (v *View) matchScopeSchemaURL(schemaURL string) bool {
 	return v.scope.SchemaURL == "" || schemaURL == v.scope.SchemaURL
 }
 
-func (v View) matchInstrumentKind(kind InstrumentKind) bool {
+func (v *View) matchInstrumentKind(kind InstrumentKind) bool {
 	return v.instrumentKind == undefinedInstrument || kind == v.instrumentKind
 }
 
-func (v View) match(i Instrument) bool {
+func (v *View) match(i Instrument) bool {
 	return v.matchName(i.Name) &&
 		v.matchScopeName(i.Scope.Name) &&
 		v.matchScopeSchemaURL(i.Scope.SchemaURL) &&
```

I believe this settles the question in #3199: it's no big deal to use structs for Views. If there's a statistical difference here, it's slight.

**Review by dmathieu — APPROVED** [review]

File: sdk/metric/benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	"go.opentelemetry.io/otel/sdk/metric/view"
+)
+
+func BenchmarkCounterAddNoAttrs(b *testing.B) {
+	ctx := context.Background()
+	rdr := NewManualReader()
+	provider := NewMeterProvider(WithReader(rdr))
+	b.ReportAllocs()
+
+	cntr, _ := provider.Meter("test").SyncInt64().Counter("hello")
```
Comment (MrAlias) [inline sdk/metric/benchmark_test.go:31]: This is repeated 7 times, can you abstract this into its own function.

File: sdk/metric/benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+	for i := 0; i < b.N; i++ {
+		cntr.Add(ctx, 1, attribute.String("", "V"), attribute.String("K", "V"))
+	}
+}
+
+func BenchmarkCounterAddManyAttrs(b *testing.B) {
```
Comment (MrAlias) [inline sdk/metric/benchmark_test.go:64]: This seems to imply that many attributes are added in one call. Maybe something like this instead.

```suggestion
func BenchmarkCounterAddSingleUseAttrs(b *testing.B) {
```

File: sdk/metric/benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+func BenchmarkCounterAddNoAttrs(b *testing.B) {
+	ctx := context.Background()
+	rdr := NewManualReader()
+	provider := NewMeterProvider(WithReader(rdr))
+	b.ReportAllocs()
+
+	cntr, _ := provider.Meter("test").SyncInt64().Counter("hello")
+
+	for i := 0; i < b.N; i++ {
```
Comment (MrAlias) [inline sdk/metric/benchmark_test.go:33]:
```suggestion
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
```

File: sdk/metric/benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		for k := 0; k < 10; k++ {
+			for j := 0; j < 10; j++ {
+				cntr.Add(ctx, 1, attribute.Int("K", j))
+			}
+			_, _ = rdr.Collect(ctx)
+		}
+	}
+}
```
Comment (MrAlias) [inline sdk/metric/benchmark_test.go:152]: This is unnecessary, it only serves to divide `b.N` by ten. It is equivalent to `BenchmarkCounterCollectTenAttrs`.

File: sdk/metric/benchmark_test.go (hunk 1 above)
Comment (jmacd) [reply sdk/metric/benchmark_test.go:31]: Done.

File: sdk/metric/benchmark_test.go (hunk 3 above)
Comment (jmacd) [reply sdk/metric/benchmark_test.go:39]: Done.

File: sdk/metric/benchmark_test.go (hunk 2 above)
Comment (jmacd) [reply sdk/metric/benchmark_test.go]: Done.

File: sdk/metric/benchmark_test.go (hunk 4 above)
Comment (jmacd) [reply sdk/metric/benchmark_test.go]: Yes, sorry -- copy and pasted mistake.

**Review by dashpole — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Pull request merged**